### DISCUSSION TITLE
Notifying missing Upgrades

### DISCUSSION THREAD

**Discussion by ceyonur** [ava-labs/subnet-evm#439] [Ideas]
Mandatory upgrades can be done without a code-release through JSON files in Subnet-EVM. This upgrade has to be coordinated for all nodes in the network. This can be a critical for indexers/API nodes as if they did not upgrade they would have no idea when they should until they get a invalid block error. We should have a mechanism to let node operators there will be a update and let them prepare for the upgrade.

From @patrick-ogrady : "I think the most alertable thing for us would be to have a metric that becomes "1" when there are upgrades to apply".

An AppRequest or AppGossip mechanism must be built to share this information. Some possible solutions:
1- We can gossip the hash of upgrades.
2- Or we can directly gossip full files and let nodes show diffs. 

Alternatively we can first send hashes and then create a apprequest/response to provide full files on demand (subjected to rate limiting).

Eventually nodes should only warn if a substantial percentage of validators tells there is an upgrade.

**Comment by aaronbuchwald** [comment]
Note: in order to gossip a hash of the `upgradeBytes` instead of the full `upgradeBytes`, we will need to establish a canonical format for the `upgradeBytes` since right now we are using JSON and since this is non-canonical, we may end up with a different hash.

imo the easiest way to do this would be to simply adding codec tags in addition to the JSON tags to the network upgrade, so that we can marshal the network upgrade and take a hash of the result.

**Reply by aaronbuchwald** [reply]
This will be applied in https://github.com/ava-labs/subnet-evm/issues/268
